### PR TITLE
Firebase messaging

### PR DESCRIPTION
**Pull request by sirambd**
**News**
- feat: Add 2 flavors (`standard`: with play services, `fdroid`: without play services)
- feat: (`standard flavor`) - Handle firebase cloud messaging if play services is available
- feat: (`standard flavor`) - Register the user device for firebase push notifications if `play service` is available
- feat: (`standard flavor`) - Process firebase push notifications in foreground, (notify threadList if needed)
- feat: (`standard flavor`) - Process firebase push notifications in background, (show notification)
- feat: (`standard flavor`) - Keep sync messages worker every 15 min
- feat: (`fdroid flavor`)  - Without play service, the sync is done every 15 min with sync worker
- feat: Notifications are displayed only when the app is in backgound

### CODE REVIEW

File: app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    val apiAvailability = GoogleApiAvailability.getInstance()
+    val errorCode = apiAvailability.isGooglePlayServicesAvailable(this)
+
+    if (errorCode == ConnectionResult.SUCCESS) {
+        checkFirebaseRegistration()
+    } else {
+        if (apiAvailability.isUserResolvableError(errorCode)) {
+            apiAvailability.makeGooglePlayServicesAvailable(this)
+        } else {
+            Toast.makeText(this, "need play services", Toast.LENGTH_SHORT).show()
```
Comment (JorisBodin) [inline app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt:45]: It must be translated

**Review by JorisBodin — CHANGES_REQUESTED** [review]

**Review by JorisBodin — APPROVED** [review]

File: app/build.gradle [hunk 2]
```diff
@@ -58,6 +59,21 @@ android {
...
+    productFlavors {
+        standard {
+            dimension 'distribution'
+            apply plugin: 'com.google.gms.google-services'
+
+            getIsDefault().set(true)
+
+            dependencies {
+                implementation 'com.google.firebase:firebase-messaging-ktx:23.1.1'
+            }
```
Comment (KevinBoulongne) [inline app/build.gradle:73]:
```suggestion
            getIsDefault().set(true)
            dependencies { implementation 'com.google.firebase:firebase-messaging-ktx:23.1.1' }
```

File: app/build.gradle [hunk 3]
```diff
@@ -58,6 +59,21 @@ android {
     kotlinOptions { jvmTarget = rootProject.ext.java_version }
 
     buildFeatures { viewBinding true }
+
+    flavorDimensions 'distribution'
+    productFlavors {
```
Comment (KevinBoulongne) [inline app/build.gradle:64]:
```suggestion

    productFlavors {
```

File: app/src/fdroid/java/com/infomaniak/mail/GplayExtensions.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+package com.infomaniak.mail
+
+import android.content.Context
+import androidx.fragment.app.FragmentActivity
+
+fun FragmentActivity.checkPlayServices() = Unit
+
+fun Context.isGooglePlayServicesNotAvailable(): Boolean = true
+
+fun Context.cancelFirebaseProcessWorks() = Unit
```
Comment (KevinBoulongne) [inline app/src/fdroid/java/com/infomaniak/mail/GplayExtensions.kt:27]:
```suggestion
fun Context.cancelFirebaseProcessWorks() = Unit

```

File: app/src/fdroid/java/com/infomaniak/mail/firebase/RegisterFirebaseBroadcastReceiver.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.mail.firebase
+
+import com.infomaniak.mail.utils.IRegisterFirebaseBroadcastReceiver
+
+class RegisterFirebaseBroadcastReceiver : IRegisterFirebaseBroadcastReceiver
```
Comment (KevinBoulongne) [inline app/src/fdroid/java/com/infomaniak/mail/firebase/RegisterFirebaseBroadcastReceiver.kt:22]:
```suggestion
class RegisterFirebaseBroadcastReceiver : IRegisterFirebaseBroadcastReceiver

```

File: app/src/main/java/com/infomaniak/mail/ui/MainActivity.kt [hunk 6]
```diff
@@ -35,15 +35,17 @@ import com.infomaniak.lib.core.networking.LiveDataNetworkStatus
...
+import com.infomaniak.mail.firebase.RegisterFirebaseBroadcastReceiver
 import com.infomaniak.mail.ui.main.menu.MenuDrawerFragment
 import com.infomaniak.mail.utils.PermissionUtils
 import com.infomaniak.mail.utils.UiUtils
-import com.infomaniak.mail.workers.SyncMessagesWorker
 import io.sentry.Breadcrumb
 import io.sentry.Sentry
 import io.sentry.SentryLevel
 
+
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainActivity.kt:48]:
```suggestion
```

File: app/src/main/java/com/infomaniak/mail/utils/FetchMessagesManager.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        when (ApiController.json.decodeFromString<ApiResponse<Any>>(exception.message!!).error?.code) {
+            ApiErrorException.ErrorCodes.FOLDER_DOES_NOT_EXIST -> Unit
+            else -> Sentry.captureException(exception)
+        }
+    }
+
+    private companion object {
+        val TAG = FetchMessagesManager::class.simpleName
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/FetchMessagesManager.kt:149]:
```suggestion
}

```

File: app/src/main/java/com/infomaniak/mail/utils/FetchMessagesManager.kt [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+        realm.close()
+    }
+
+    private suspend fun Thread.showNotification(
+        userId: Int,
+        mailbox: Mailbox,
+        unReadThreadsCount: Int,
+        realm: Realm,
+        okHttpClient: OkHttpClient
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/FetchMessagesManager.kt:83]:
```suggestion
        okHttpClient: OkHttpClient,
```

File: app/src/main/java/com/infomaniak/mail/utils/FetchMessagesManager.kt [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        val preview = message.body?.value?.ifBlank { null }
+            ?.let { "\n${MessageBodyUtils.splitBodyAndQuote(it).messageBody.htmlToText().trim()}" }
+            ?: message.preview.ifBlank { null }?.let { "\n${it.trim()}" }
+            ?: ""
+        val formattedPreview = preview.replace("\\n+\\s*".toRegex(), "\n") // Ignore multiple/start whitespaces
+        val description = "$subject$formattedPreview"
+
+        // Show message notification
+        showNotification(subject, false, message.sender.displayedName(context), description)
+        // Show group summary notification
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/FetchMessagesManager.kt:130]:
```suggestion

        // Show group summary notification
```

File: app/src/main/java/com/infomaniak/mail/utils/IRegisterFirebaseBroadcastReceiver.kt [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+package com.infomaniak.mail.utils
+
+import androidx.fragment.app.FragmentActivity
+import com.infomaniak.mail.ui.MainViewModel
+
+interface IRegisterFirebaseBroadcastReceiver {
+
+    fun initFirebaseBroadcastReceiver(activity: FragmentActivity, mainViewModel: MainViewModel) = Unit
+
+}
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/IRegisterFirebaseBroadcastReceiver.kt:27]:
```suggestion
}

```

File: app/src/main/java/com/infomaniak/mail/workers/SyncMailboxesWorker.kt [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+        Result.success()
+    }
+
+    override fun onFinish() {
+        mailboxInfoRealm.close()
+    }
+
+    companion object {
+        /** To support the old services, we do not change the name */
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/workers/SyncMailboxesWorker.kt:57]:
```suggestion

        /** To support the old services, we do not change the name */
```

File: app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt (hunk 1 above)
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt:45]:
```suggestion
            // TODO: Translate this
            Toast.makeText(this, "Need Play Services", Toast.LENGTH_SHORT).show()
```

File: app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+private fun FragmentActivity.checkFirebaseRegistration() {
+    lifecycleScope.launchWhenCreated {
+        checkFirebaseRegistration(this@checkFirebaseRegistration)
+    }
+}
+
+private suspend fun checkFirebaseRegistration(context: Context) = withContext(Dispatchers.IO) {
+    val localSettings = LocalSettings.getInstance(context)
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt:66]:
```suggestion

    val localSettings = LocalSettings.getInstance(context)
```

File: app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        val token = task.result
+        Log.d("firebase", "checkFirebaseRegistration: token is different ${token != localSettings.firebaseToken}")
+        if (token != localSettings.firebaseToken) localSettings.clearRegisteredFirebaseUsers()
+
+        localSettings.firebaseToken = token
+        RegisterUserDeviceWorker.scheduleWork(context)
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/GplayExtensions.kt:86]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/FirebaseApiRepository.kt [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import okhttp3.OkHttpClient
+
+object FirebaseApiRepository {
+
+    private const val registerDevice = "$INFOMANIAK_API/devices/register"
+
+    fun registerForNotifications(registrationInfos: RegistrationInfos, okHttpClient: OkHttpClient): ApiResponse<Boolean> {
+        return ApiRepository.callApi(registerDevice, POST, registrationInfos, okHttpClient)
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/FirebaseApiRepository.kt:33]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/KMailFirebaseMessagingService.kt [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        Log.i(TAG, "processMessageInForeground: called")
+        if (AccountUtils.currentUserId == userId && AccountUtils.currentMailboxId == mailboxId) {
+            LocalBroadcastManager.getInstance(this).sendBroadcast(Intent(ACTION_MESSAGE_RECEIVED))
+        }
+    }
+
+    private fun processMessageInBackground(userId: Int, mailboxId: Int, messageUid: String) {
+        Log.i(TAG, "processMessageInBackground: called")
+        MailboxController.getMailbox(userId, mailboxId, realmMailboxInfo)?.let { mailbox ->
+            // Ignore if the mailbox notification channel is blocked
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/KMailFirebaseMessagingService.kt:77]:
```suggestion
            // Ignore if the Mailbox notification channel is blocked
```

File: app/src/standard/java/com/infomaniak/mail/firebase/KMailFirebaseMessagingService.kt [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    override fun onDestroy() {
+        realmMailboxInfo.close()
+        super.onDestroy()
+    }
+
+    companion object {
+        private const val TAG = "KMailFirebaseMessagingService"
+        const val ACTION_MESSAGE_RECEIVED = "action_message_received"
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/KMailFirebaseMessagingService.kt:93]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/ProcessMessageNotificationsWorker.kt [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+            WorkManager.getInstance(context).enqueueUniqueWork(workName, ExistingWorkPolicy.APPEND_OR_REPLACE, workRequest)
+        }
+
+        fun cancelWorks(context: Context) {
+            WorkManager.getInstance(context).cancelAllWorkByTag(TAG)
+        }
+
+        private fun workName(userId: Int, mailboxId: Int) = "${userId}_$mailboxId"
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/ProcessMessageNotificationsWorker.kt:89]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/RegisterFirebaseBroadcastReceiver.kt [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+    override fun onResume(owner: LifecycleOwner) {
+        val intentFilter = IntentFilter(KMailFirebaseMessagingService.ACTION_MESSAGE_RECEIVED)
+        localBroadcastManager.registerReceiver(firebaseBroadcastReceiver, intentFilter)
+    }
+
+    override fun onPause(owner: LifecycleOwner) {
+        localBroadcastManager.unregisterReceiver(firebaseBroadcastReceiver)
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/RegisterFirebaseBroadcastReceiver.kt:59]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/RegisterUserDeviceWorker.kt [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            Log.i(TAG, "work scheduled")
+
+            val workRequest = OneTimeWorkRequestBuilder<RegisterUserDeviceWorker>()
+                .setConstraints(Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build())
+                .build()
+
+            WorkManager.getInstance(context).enqueueUniqueWork(TAG, ExistingWorkPolicy.REPLACE, workRequest)
+        }
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/RegisterUserDeviceWorker.kt:78]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/RegistrationInfos.kt [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+
+    private companion object {
+        private const val OS_NAME = "android"
+        private val DEVICE_MODEL = android.os.Build.MODEL
+
+        private fun getDeviceName(contentResolver: ContentResolver): String {
+            return Settings.Global.getString(contentResolver, "device_name")
+        }
+    }
+}
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/RegistrationInfos.kt:41]:
```suggestion
}

```

File: app/src/standard/java/com/infomaniak/mail/firebase/RegistrationInfos.kt [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    private val model: String = DEVICE_MODEL,
+) {
+
+    constructor(context: Context, token: String) : this(token = token, name = getDeviceName(context.contentResolver))
+
+    private companion object {
+        private const val OS_NAME = "android"
+        private val DEVICE_MODEL = android.os.Build.MODEL
+
+        private fun getDeviceName(contentResolver: ContentResolver): String {
```
Comment (KevinBoulongne) [inline app/src/standard/java/com/infomaniak/mail/firebase/RegistrationInfos.kt:37]:
```suggestion
        const val OS_NAME = "android"
        val DEVICE_MODEL: String = android.os.Build.MODEL

        fun getDeviceName(contentResolver: ContentResolver): String {
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**